aiohttp/web_request.py

```diff
@@ -253,19 +257,31 @@ def rel_url(self) -> URL:
 
     # MutableMapping API
 
-    def __getitem__(self, key: str) -> Any:
+    @overload  # type: ignore[override]
+    def __getitem__(self, key: RequestKey[_T]) -> _T: ...
+
+    @overload
+    def __getitem__(self, key: str) -> Any: ...
+
+    def __getitem__(self, key: str | RequestKey[_T]) -> Any:
         return self._state[key]
 
-    def __setitem__(self, key: str, value: Any) -> None:
+    @overload  # type: ignore[override]
+    def __setitem__(self, key: RequestKey[_T], value: _T) -> None: ...
+
+    @overload
+    def __setitem__(self, key: str, value: Any) -> None: ...
+
+    def __setitem__(self, key: str | RequestKey[_T], value: Any) -> None:
         self._state[key] = value
```

Comment:
We should probably put the same warning in:
```suggestion
        if not isinstance(key, RequestKey):
            warnings.warn(
                "It is recommended to use web.RequestKey instances for keys.\n"
                + "https://docs.aiohttp.org/en/stable/web_advanced.html"
                + "#application-s-config",
                category=NotAppKeyWarning,
                stacklevel=2,
            )
        self._state[key] = value
```

We can probably just leave the warning as NotAppKeyWarning for backwards compatibility..

Reply:
Added the warnings, but, to reiterate, there'll probably be a lot of them, if they're enabled.

It might be better to fire them once (add a flag attribute in each class or `Application` instance) or use some kind of debouncer to limit their frequency.

Reply:
> Added the warnings, but, to reiterate, there'll probably be a lot of them, if they're enabled.

Yeah, you're probably right. Might need to tweak it then. I was also thinking it could be a minor performance issue, so maybe we only keep these warnings in for 1 release to help notify users, and then remove it again in the next release..

Reply:
Added a debouncer class to limit the number of warnings raised ([7df97de](https://github.com/aio-libs/aiohttp/pull/11766/commits/7df97de12cafa888a54392047a367ba06d528903), [f2bea0b](https://github.com/aio-libs/aiohttp/pull/11766/commits/f2bea0b4c52cef68d5cbb4543bdde23c5e9ef587)).

At the same time, I don't quite understand, if it needs to be thread-safe, given that its instances in request & response classes' attributes are global. If it is, we could add a critical section into `DebounceContextManager.__enter__` to avoid potential races, when updating/counting the number of entries done.